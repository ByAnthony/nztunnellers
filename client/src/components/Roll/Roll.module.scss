@use "src/index.scss";

.roll {
  @include index.page-setup;
}

.letter-container {
  width: 150px;
  border-radius: 5px;
  background-color: rgb(173, 161, 120);
}

.letter-title {
  padding-top: 15px;
  font-size: 125px;
  font-stretch: 85%;
  font-weight: 700;
  color: rgb(24, 26, 27);
  text-align: center;

  /* Mobile
  margin: 100px 0;
  font-size: 125px;
  font-stretch: 85%;
  font-weight: 700;
  */
}

.tunneller-group {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  width: 100%;
}

.tunneller {
  margin: 75px 0;
  text-align: left;
  flex-basis: 50%;
}

.surname {
  margin: 0;
  font-size: 50px;
  font-stretch: 85%;
  font-weight: 700;
  letter-spacing: 1px;

  /* Mobile
  margin: 0;
  font-size: 30px;
  font-stretch: 85%;
  font-weight: 700;
  letter-spacing: 1px;
  text-transform: uppercase;
  */
}

.forename {
  margin: 0;
  font-size: 35px;
  font-weight: 300;
  font-stretch: 85%;
  letter-spacing: 1px;

  /* Mobile
  margin: 0;
  font-size: 25px;
  font-weight: 300;
  font-stretch: 85%;
  letter-spacing: 1px;
  */
}

.serial {
  margin-top: 25px;
  font-size: 20px;
  font-stretch: 85%;
  font-weight: 200;
  letter-spacing: 2px;

  /* Mobile
  margin-top: 15px;
  font-size: 16px;
  font-stretch: 85%;
  font-weight: 200;
  letter-spacing: 2px;
  */
}
