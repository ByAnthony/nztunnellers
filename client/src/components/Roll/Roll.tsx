import { Tunneller } from '../../types';
import { RollTunneller } from '../RollTunneller/RollTunneller';
import STYLES from './Roll.module.scss';

type Tunnellers = {
    tunnellers: Record<string, Tunneller[]> | never[];
    filterByLetter: string;
};

export const Roll = ({ tunnellers, filterByLetter }: Tunnellers) => {

    const tunnellersList = Object.entries(tunnellers);

    const isFilteredByLetter = (letter: string) => {
        return letter === '' ? tunnellersList : tunnellersList.filter((key) => key.includes(letter))
    }

    const companyRoll = isFilteredByLetter(filterByLetter)
        .map(([key, listOfTunnellers]) => (
            <div id={`letter-${key}`} key={key}>
                <div className={STYLES['letter-container']}>
                    <h2 className={STYLES['letter-title']} key={key}>{key}</h2>
                </div>
                <div className={STYLES['tunnellers-container']}>
                    <RollTunneller listOfTunnellers={listOfTunnellers} />
                </div>
            </div>
        ));

    return(
        <div className={STYLES.roll}>
          {companyRoll}
        </div>
    );

};
