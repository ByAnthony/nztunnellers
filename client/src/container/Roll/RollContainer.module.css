.alphabet {
    margin: 0 auto;
    width: 75%;
}

.button-container {
    display: grid;
    grid-template-columns: repeat(auto-fill, 50px);
    column-gap: 50px;
    row-gap: 25px;
    margin: 75px 0;
    width: 100%;
}

button {
    padding: 10px 0 7px 0;
    font-family: 'Mona Sans';
    font-size: 30px;
    font-stretch: 85%;
    font-weight: 600;
    color: rgb(173,161,120);
    text-align: center;
    text-decoration: none;
    background-color: transparent;
    border: none;
    cursor: pointer;
}

button:hover {
    color: rgb(24,26,27);
    background-color: rgb(173,161,120);
    border-radius: 3px;
}
