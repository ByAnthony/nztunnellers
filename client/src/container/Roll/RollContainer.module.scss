@use 'src/index.scss';

.alphabet-container {
  @include index.component-setup;
}

.alphabet {
  width: 100%;
  margin: 50px 0;
  display: grid;
  grid-template-columns: repeat(auto-fill, 30px);
  justify-content: space-between;
  grid-gap: 20px;

  @include index.tablet {
    margin: 75px 0;
    grid-template-columns: repeat(auto-fill, 40px);
    grid-gap: 25px;
  }

  @include index.desktop {
    margin: 100px 0;
    grid-template-columns: repeat(auto-fill, 50px);
    grid-gap: 30px;
  }
}

.alphabet-letter {
  padding: 10px 0 5px 0;
  font-family: index.$main-font;
  font-size: 16px;
  font-stretch: 85%;
  font-weight: 600;
  color: index.$tertiary-color;
  text-align: center;
  text-decoration: none;
  background-color: transparent;
  border: none;
  cursor: pointer;

  &:hover {
    color: index.$primary-color;
    background-color: index.$tertiary-color;
    border-radius: 3px;
  }

  @include index.tablet {
    font-size: 26px;
  }

  @include index.desktop {
    font-size: 30px;
  }
}
