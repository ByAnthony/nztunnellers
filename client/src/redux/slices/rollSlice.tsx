import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { Details } from '../../types/roll';
import { Profile } from '../../types/tunneller';

export const tunnellersApi = createApi({
  reducerPath: 'tunnellersApi',
  baseQuery: fetchBaseQuery({
    baseUrl: 'http://localhost:5000/',
  }),
  endpoints: (builder) => ({
    getAllTunnellers: builder.query<Record<string, Array<Details>>, void>({
      query: () => ({
        url: 'roll/',
      }),
    }),
    getTunnellerById: builder.query<Profile, number>({
      query: (id: number) => `roll/${id}`,
    }),
  }),
});

export const { useGetAllTunnellersQuery, useGetTunnellerByIdQuery } = tunnellersApi;
