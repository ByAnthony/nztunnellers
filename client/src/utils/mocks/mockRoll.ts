import { mockId } from './mockProfile';

const johnDoe = {
  forename: 'John',
  surname: 'Doe',
};

const alexanderDriscott = {
  forename: 'Alexander',
  surname: 'Driscott',
};

const williamRight = {
  forename: 'William',
  surname: 'Right',
};

const robertLang = {
  forename: 'Robert',
  surname: 'Lang',
};

export const mockDetails = {
  id: mockId,
  name: johnDoe,
  birth: '1886',
  death: '1952',
};

export const mockRoll = {
  D: [
    mockDetails,
    {
      ...mockDetails,
      id: 1,
      name: alexanderDriscott,
    },
  ],
  L: [
    {
      ...mockDetails,
      id: 2,
      name: robertLang,
    },
  ],
  R: [
    {
      ...mockDetails,
      id: 3,
      name: williamRight,
    },
  ],
};
