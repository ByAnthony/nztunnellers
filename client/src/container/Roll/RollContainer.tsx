import { useState } from 'react';
import { Roll } from '../../components/Roll/Roll';
import STYLES from './RollContainer.module.scss';
import { useSelector } from 'react-redux';

export const RollContainer = () => {

    // const [tunnellers, setAllTunnellers] = useState([]);
    const [filterByLetter, setFilterByLetter] = useState('');

    // useEffect(() => {
    //     RollService.getRollSortedByAlphabet()
    //     .then(tunnellers => setAllTunnellers(tunnellers));
    // }, []);

    const tunnellers = useSelector((state: any) => state.roll.value)
    const letters = Object.keys(tunnellers);

    return(
        <>
            <div className={STYLES['alphabet-container']}>
                <h1>Company Roll</h1>
                <div className={STYLES['alphabet']}>
                    {letters.map(letter => <button key={letter} className={STYLES['alphabet-letter']} onClick={() => setFilterByLetter(letter)}>{letter}</button>)}
                    <button key='All' className={STYLES['alphabet-letter']} onClick={() => setFilterByLetter('')}>All</button>
                </div>
            </div>
            <Roll tunnellers={tunnellers} filterByLetter={filterByLetter} />
        </>
    );

};
