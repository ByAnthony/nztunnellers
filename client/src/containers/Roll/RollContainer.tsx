import { useState } from 'react';
import { Roll } from '../../components/Roll/Roll';
import STYLES from './RollContainer.module.scss';
import { useGetAllTunnellersQuery } from '../../redux/slices/rollSlice';

export function RollContainer() {
  const [filterByLetter, setFilterByLetter] = useState('');

  const { data, isLoading, isSuccess } = useGetAllTunnellersQuery();

  if (data) {
    const letters = Object.keys(data);
    return (
      <>
        { isLoading }
        { isSuccess && (
        <>
          <div className={STYLES['alphabet-container']}>
            <h1>Company Roll</h1>
            <div className={STYLES.alphabet}>
              {letters.map((letter) => <button type="button" key={letter} className={STYLES['alphabet-letter']} onClick={() => setFilterByLetter(letter)}>{letter}</button>)}
              <button type="button" key="All" className={STYLES['alphabet-letter']} onClick={() => setFilterByLetter('')}>All</button>
            </div>
          </div>
          <Roll tunnellers={data} filterByLetter={filterByLetter} />
        </>
        )}
      </>
    );
  }
  return null;
}
