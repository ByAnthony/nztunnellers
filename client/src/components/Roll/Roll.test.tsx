import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { mockRoll } from '../../utils/mocks/mockRoll';
import { useGetAllTunnellersQuery } from '../../redux/slices/rollSlice';
import { Roll } from './Roll';

jest.mock('../../redux/slices/rollSlice', () => ({
  useGetAllTunnellersQuery: jest.fn(),
}));

const component = (
  <MemoryRouter>
    <Roll />
  </MemoryRouter>
);

test('renders roll when data is available', () => {
  (useGetAllTunnellersQuery as jest.Mock).mockReturnValue({
    data: mockRoll,
    error: null,
    isLoading: false,
    isSuccess: true,
  });

  const { asFragment } = render(component);

  expect(asFragment()).toMatchSnapshot();

  const buttonD = screen.getByLabelText('Filter names by the letter D');
  const buttonL = screen.getByLabelText('Filter names by the letter L');
  const buttonR = screen.getByLabelText('Filter names by the letter R');
  expect(buttonD).toBeInTheDocument();
  expect(buttonD).toHaveTextContent('D');
  expect(buttonL).toBeInTheDocument();
  expect(buttonL).toHaveTextContent('L');
  expect(buttonR).toBeInTheDocument();
  expect(buttonR).toHaveTextContent('R');

  const buttonAll = screen.getByLabelText('Remove the filter by name');
  expect(buttonAll).toBeInTheDocument();
  expect(buttonAll).toHaveTextContent('All');

  const titleD = screen.getByLabelText('Letter D');
  const titleL = screen.getByLabelText('Letter L');
  const titleR = screen.getByLabelText('Letter R');
  expect(titleD).toHaveClass('title');
  expect(titleD).toHaveTextContent('D');
  expect(titleD).toBeInTheDocument();
  expect(titleL).toHaveClass('title');
  expect(titleL).toHaveTextContent('L');
  expect(titleL).toBeInTheDocument();
  expect(titleR).toHaveClass('title');
  expect(titleR).toHaveTextContent('R');
  expect(titleR).toBeInTheDocument();

  expect(screen.getByText('John')).toBeInTheDocument();
  expect(screen.getByText('Doe')).toBeInTheDocument();

  expect(screen.getByText('Alexander')).toBeInTheDocument();
  expect(screen.getByText('Driscott')).toBeInTheDocument();

  expect(screen.getByText('Robert')).toBeInTheDocument();
  expect(screen.getByText('Lang')).toBeInTheDocument();

  expect(screen.getByText('William')).toBeInTheDocument();
  expect(screen.getByText('Right')).toBeInTheDocument();
});

describe('Filter', () => {
  test('should filter by name', () => {
    (useGetAllTunnellersQuery as jest.Mock).mockReturnValue({
      data: mockRoll,
      error: null,
      isLoading: false,
      isSuccess: true,
    });

    render(component);

    const buttonD = screen.getByLabelText('Filter names by the letter D');
    fireEvent.click(buttonD);
    expect(screen.getByLabelText('Letter D')).toBeInTheDocument();
    expect(screen.queryByLabelText('Letter L')).not.toBeInTheDocument();
    expect(screen.queryByLabelText('Letter R')).not.toBeInTheDocument();

    expect(screen.getByText('John')).toBeInTheDocument();
    expect(screen.getByText('Doe')).toBeInTheDocument();

    expect(screen.getByText('Alexander')).toBeInTheDocument();
    expect(screen.getByText('Driscott')).toBeInTheDocument();

    expect(screen.queryByText('Robert')).not.toBeInTheDocument();
    expect(screen.queryByText('Lang')).not.toBeInTheDocument();

    expect(screen.queryByText('William')).not.toBeInTheDocument();
    expect(screen.queryByText('Right')).not.toBeInTheDocument();
  });

  test('should remove filter by name', () => {
    (useGetAllTunnellersQuery as jest.Mock).mockReturnValue({
      data: mockRoll,
      error: null,
      isLoading: false,
      isSuccess: true,
    });

    render(component);

    const buttonD = screen.getByLabelText('Filter names by the letter D');
    const buttonAll = screen.getByLabelText('Remove the filter by name');

    fireEvent.click(buttonD);

    expect(screen.getByLabelText('Letter D')).toBeInTheDocument();
    expect(screen.queryByLabelText('Letter L')).not.toBeInTheDocument();
    expect(screen.queryByLabelText('Letter R')).not.toBeInTheDocument();

    fireEvent.click(buttonAll);

    expect(screen.getByLabelText('Letter D')).toBeInTheDocument();
    expect(screen.getByLabelText('Letter L')).toBeInTheDocument();
    expect(screen.getByLabelText('Letter R')).toBeInTheDocument();
  });
});

test('does not render roll when data is undefined', () => {
  (useGetAllTunnellersQuery as jest.Mock).mockReturnValue({
    data: undefined,
    error: null,
    isLoading: false,
    isSuccess: true,
  });

  const { container } = render(<Roll />);

  expect(container).toBeEmptyDOMElement();
});
