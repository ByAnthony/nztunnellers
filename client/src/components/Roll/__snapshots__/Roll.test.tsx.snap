// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders roll when data is available 1`] = `
<DocumentFragment>
  <div
    class="menu "
  >
    <a
      aria-label="Go to the Homepage"
      class="logo"
      href="/"
    >
      <img
        alt=""
        src="/nzt_logo.png"
      />
    </a>
    <div
      class="search-form"
    >
      <form>
        <input
          id="search"
          name="search"
          placeholder="Search for a Tunneller"
          type="text"
        />
      </form>
      <button
        class="search-form-button"
        type="submit"
      >
        <img
          alt=""
          height="18"
          src="/searching_hover.png"
        />
      </button>
    </div>
  </div>
  <div
    class="container"
  >
    <div
      class="header"
    >
      <div
        class="main-title"
      >
        <h1>
          <span
            class="title-line-1"
          >
            The New Zealand
          </span>
          <span
            class="title-line-2"
          >
            Tunnellers
          </span>
        </h1>
      </div>
    </div>
    <div
      class="roll-container"
    >
      <div
        class="alphabet"
      >
        <button
          aria-label="Filter names by the letter D"
          class="letter"
          type="button"
        >
          D
        </button>
        <button
          aria-label="Filter names by the letter L"
          class="letter"
          type="button"
        >
          L
        </button>
        <button
          aria-label="Filter names by the letter R"
          class="letter"
          type="button"
        >
          R
        </button>
        <button
          aria-label="Remove the filter by name"
          class="letter"
          type="button"
        >
          All
        </button>
      </div>
      <div
        class="roll"
      >
        <div
          id="letter-D"
        >
          <div
            class="letter-container"
          >
            <h2
              aria-label="Letter D"
              class="title"
            >
              D
            </h2>
          </div>
          <div
            class="tunnellers-container"
          >
            <a
              href="/tunnellers/26"
            >
              <div
                class="tunneller"
              >
                <div>
                  <p
                    class="forename"
                  >
                    John
                  </p>
                  <p
                    class="surname"
                  >
                    Doe
                  </p>
                  <p
                    class="dates"
                  >
                    1886-1952
                  </p>
                </div>
                <div
                  class="arrow"
                >
                  →
                </div>
              </div>
            </a>
            <a
              href="/tunnellers/1"
            >
              <div
                class="tunneller"
              >
                <div>
                  <p
                    class="forename"
                  >
                    Alexander
                  </p>
                  <p
                    class="surname"
                  >
                    Driscott
                  </p>
                  <p
                    class="dates"
                  >
                    1886-1952
                  </p>
                </div>
                <div
                  class="arrow"
                >
                  →
                </div>
              </div>
            </a>
          </div>
        </div>
        <div
          id="letter-L"
        >
          <div
            class="letter-container"
          >
            <h2
              aria-label="Letter L"
              class="title"
            >
              L
            </h2>
          </div>
          <div
            class="tunnellers-container"
          >
            <a
              href="/tunnellers/2"
            >
              <div
                class="tunneller"
              >
                <div>
                  <p
                    class="forename"
                  >
                    Robert
                  </p>
                  <p
                    class="surname"
                  >
                    Lang
                  </p>
                  <p
                    class="dates"
                  >
                    1886-1952
                  </p>
                </div>
                <div
                  class="arrow"
                >
                  →
                </div>
              </div>
            </a>
          </div>
        </div>
        <div
          id="letter-R"
        >
          <div
            class="letter-container"
          >
            <h2
              aria-label="Letter R"
              class="title"
            >
              R
            </h2>
          </div>
          <div
            class="tunnellers-container"
          >
            <a
              href="/tunnellers/3"
            >
              <div
                class="tunneller"
              >
                <div>
                  <p
                    class="forename"
                  >
                    William
                  </p>
                  <p
                    class="surname"
                  >
                    Right
                  </p>
                  <p
                    class="dates"
                  >
                    1886-1952
                  </p>
                </div>
                <div
                  class="arrow"
                >
                  →
                </div>
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="footer"
  >
    <div
      class="map"
    >
      <div
        class="links"
      >
        <div
          class="map-link"
        >
          <a
            href="/#history"
          >
            History
          </a>
        </div>
        <div
          class="map-link"
        >
          <a
            href="/tunnellers"
          >
            Tunnellers
          </a>
        </div>
        <div
          class="map-link"
        >
          <a
            href="/about-us"
          >
            About Us
          </a>
        </div>
      </div>
      <button
        aria-label="Go back to the top of the page"
        class="scroll-top"
        type="button"
      >
        ↑
      </button>
    </div>
    <div
      class="support"
    >
      <div>
        <a
          aria-label="Go to The Artois University website"
          href="https://www.univ-artois.fr/artois-university"
        >
          <img
            alt="Artois University homepage"
            src="/images/support/logo-univ-artois-blanc_0.png"
          />
        </a>
      </div>
      <div>
        <a
          aria-label="Go to The Institute for Strategic Research website"
          href="https://www.irsem.fr/en/"
        >
          <img
            alt="Institute for Strategic Research homepage"
            src="/images/support/irsem-white.png"
          />
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
